core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -124,15 +134,73 @@ object DumpLogSegments {
         println("offset: %d position: %d".format(entry.offset + index.baseOffset, entry.position))
     }
   }
-  
+
+  private trait MessageParser[K, V] {
+    def parse(message: Message): (Option[K], Option[V])
+  }
+
+  private class DecoderMessageParser[K, V](keyDecoder: Decoder[K], valueDecoder: Decoder[V]) extends MessageParser[K, V] {
+    override def parse(message: Message): (Option[K], Option[V]) = {
+      if (message.isNull) {
+        (None, None)
+      } else {
+        val key = if (message.hasKey)
+          Some(keyDecoder.fromBytes(Utils.readBytes(message.key)))
+        else
+          None
+
+        val payload = Some(valueDecoder.fromBytes(Utils.readBytes(message.payload)))
+
+        (key, payload)
+      }
+    }
+  }
+
+  private class OffsetsMessageParser extends MessageParser[String, String] {
+    private def parseOffsets(offsetKey: OffsetKey, payload: ByteBuffer) = {
+      val group = offsetKey.key.group
+      val (topic, partition)  = offsetKey.key.topicPartition.asTuple
+      val offset = GroupMetadataManager.readOffsetMessageValue(payload)
+      (Some(s"offset::${group}:${topic}:${partition}"), Some(s"${offset.offset}:${offset.metadata}"))
+    }
+
+    private def parseGroupMetadata(groupMetadataKey: GroupMetadataKey, payload: ByteBuffer) = {
+      val groupId = groupMetadataKey.key
+      val group = GroupMetadataManager.readGroupMessageValue(groupId, payload)
+      val protocolType = group.protocolType
+
+      val assignment = group.allMemberMetadata.map { member =>
+        if (protocolType == ConsumerProtocol.PROTOCOL_TYPE) {
+          val partitions = ConsumerProtocol.deserializeAssignment(ByteBuffer.wrap(member.assignment)).partitions()
+          s"${member.memberId}=${partitions}"
+        } else {
+          s"${member.memberId}=${String.format("%x", BigInt(1, member.assignment))}"
+        }
+      }.mkString("{", ",", "}")
+
+      (Some(s"metadata::${groupId}"), Some(s"${protocolType}:${group.protocol}:${group.generationId}:${assignment}"))
+    }
+
+    override def parse(message: Message): (Option[String], Option[String]) = {
+      if (message.isNull || !message.hasKey) {
+        (None, None)
```

Comment:
Instead of returning None, it may be better printing an error and either exit the program or skip this message.